app/components/shared/Duration.js

```diff
@@ -23,67 +46,99 @@ class Duration extends React.PureComponent {
+    updateFrequency: second.bind(1),
     tabularNumerals: true
   };
 
-  state = {
-    value: ''
-  };
+  _timeout: TimeoutID;
 
-  updateTime() {
-    const { from, to, format } = this.props;
+  state = {};
+
+  tick() {
+    const { from, to } = this.props;
 
     this.setState({
-      value: getDurationString(from, to, format)
+      seconds: getDuration(from, to).seconds()
+    }, () => {
+      this.maybeScheduleTick();
     });
   }
 
   componentDidMount() {
-    this.maybeSetInterval(this.props.updateFrequency);
+    this.maybeScheduleTick();
   }
 
-  maybeSetInterval(updateFrequency) {
-    if (updateFrequency > 0) {
-      this._interval = setInterval(() => this.updateTime(), updateFrequency);
+  maybeScheduleTick() {
+    const { from, to, updateFrequency } = this.props;
+
+    // We only want to schedule ticks if our duration is indeterminate,
+    // and our update frequency isn't zero
+    if (!(from && to) && typeof updateFrequency == 'number' && updateFrequency > 0) {
+      this._timeout = setTimeout(() => this.tick(), updateFrequency);
     }
-    this.updateTime();
   }
 
-  maybeClearInterval() {
-    if (this._interval) {
-      clearInterval(this._interval);
+  maybeCancelTick() {
+    if (this._timeout) {
+      clearTimeout(this._timeout);
     }
   }
 
   componentWillUnmount() {
-    this.maybeClearInterval();
+    this.maybeCancelTick();
   }
 
-  componentWillReceiveProps(nextProps) {
-    const { from, to, format, updateFrequency } = nextProps;
+  static getDerivedStateFromProps(nextProps, currentState) {
+    const seconds = getDuration(nextProps.from, nextProps.to).seconds();
 
-    if (updateFrequency !== this.props.updateFrequency) {
-      this.maybeClearInterval();
-      this.maybeSetInterval(updateFrequency);
+    if (seconds !== currentState.seconds) {
+      return { seconds };
     }
 
-    this.setState({
-      value: getDurationString(from, to, format)
-    });
+    return null;
+  }
+
+  componentDidUpdate(prevProps) {
+    const { updateFrequency, to } = this.props;
+
+    if (updateFrequency !== prevProps.updateFrequency || to !== prevProps.to) {
+      this.maybeCancelTick();
+      this.maybeScheduleTick();
+    }
+  }
+
+  handleVisibilityChange = (visible) => {
+    this.maybeCancelTick();
+
+    if (visible) {
+      this.tick();
+    }
   }
 
   render() {
-    const { state: { value }, props: { className, tabularNumerals } } = this;
     const spanClassName = classNames(
-      className,
-      { 'tabular-numerals': tabularNumerals }
+      this.props.className,
+      { 'tabular-numerals': this.props.tabularNumerals }
     );
 
+    const durationString = getDurationString(this.state.seconds, this.props.format);
+
+    if (this.props.to) {
+      return (
+        <span className={spanClassName}>
+          {durationString}
+        </span>
+      );
+    }
+
     return (
-      <span className={spanClassName}>
-        {value}
-      </span>
+      <VisibilityObserver
```

Comment:
This is neat!

Reply:
Yeah! I was super glad to see there was an implementation and polyfill for this tech! It’s solid!